src/Compilers/Core/Portable/SourceGeneration/GeneratorDriver.cs

```diff
@@ -183,6 +122,153 @@ public GeneratorDriver AddAdditionalTexts(ImmutableArray<AdditionalText> additio
             return FromState(newState);
         }
 
+        public GeneratorDriverRunResult GetRunResult()
+        {
+            var trees = ArrayBuilder<SyntaxTree>.GetInstance();
+            var diagnostics = ArrayBuilder<Diagnostic>.GetInstance();
+            var results = ArrayBuilder<GeneratorRunResult>.GetInstance(_state.Generators.Length);
+            for (int i = 0; i < _state.Generators.Length; i++)
+            {
+                var generator = _state.Generators[i];
+                var generatorState = _state.GeneratorStates[i];
+
+                results.Add(new GeneratorRunResult(generator, getSourceResults(generatorState), generatorState.Diagnostics, generatorState.Exception));
+                trees.AddRange(generatorState.Trees);
+                diagnostics.AddRange(generatorState.Diagnostics);
+            }
+
+            return new GeneratorDriverRunResult(results.ToImmutableAndFree(), trees.ToImmutableAndFree(), diagnostics.ToImmutableAndFree());
+
+            static ImmutableArray<GeneratedSourceResult> getSourceResults(GeneratorState generatorState)
+            {
+                ArrayBuilder<GeneratedSourceResult> sourceResults = ArrayBuilder<GeneratedSourceResult>.GetInstance(generatorState.SourceTexts.Length);
+                for (int i = 0; i < generatorState.SourceTexts.Length; i++)
+                {
+                    var sourceText = generatorState.SourceTexts[i];
+                    var tree = generatorState.Trees[i];
+
+                    sourceResults.Add(new GeneratedSourceResult(tree, sourceText.Text, sourceText.HintName));
+                }
+                return sourceResults.ToImmutableAndFree();
```

Comment:
>return [](start = 16, length = 6)

```
return generatorState.SourceTexts.ZipAsArray(
    generatorState.Trees,
    (sourceText, tree) => new GeneratedSourceResult(tree, sourceText.Text, sourceText.HintName));
```
(Sorry, if I'd recognized that earlier, I wouldn't have suggested creating a local function.)

Reply:
Ooh I didn't know about that one :)

---
In reply to: [466531602](https://github.com/dotnet/roslyn/pull/46296#discussion_r466531602) [](ancestors = 466531602)